### ISSUE TITLE
NemotronH LoRA: default target_modules only cover 7% of layers, silent failure

### ISSUE THREAD

**Issue by davidnichols-ops** [huggingface/peft#3554]
## Problem

PEFT's default LoRA target modules for `nemotron_h` (added in PR #3289) only target attention projections (`q_proj`, `k_proj`, `v_proj`, `o_proj`). For NemotronH's hybrid Mamba-2 + Attention architecture, this means LoRA only touches **4 out of 56 layers (7.1%)** — the other 52 layers are Mamba SSM layers with different parameter names (`in_proj`, `out_proj`, `A_log`, `D`, `dt`, `x_proj`, `z_proj`).

Worse, PEFT **silently skips** target module names that don't exist in the model. There's no warning that the configured target modules only matched a small fraction of layers. The user sees training "complete" with loss not improving, and has no idea why.

## Reproduction

We ran DPO (LoRA) on `nvidia/NVIDIA-Nemotron-Nano-9B-v2` with the default PEFT target modules:

```python
from peft import LoraConfig
config = LoraConfig(
    target_modules=["q_proj", "k_proj", "v_proj", "o_proj"],  # PEFT default for nemotron_h
    r=16, lora_alpha=32, lora_dropout=0.1,
)
```

Results:
- **Training completed** (25/25 steps, 12 minutes)
- **Loss stayed at 0.6931 (ln(2))** for all 25 steps — zero learning
- **HumanEval: 0.0% (0/164)** — the LoRA adapter produced no behavioral change

The Nemotron-Nano-9B-v2 has 56 layers with this hybrid pattern:
```
M-M-M-MM-M-M-M*-M-M-M*-M-M-M-M*-M-M-M-M*-M-MM-M-M-M-M-M-
```
Where `M` = Mamba, `*` = Attention. Only 4 layers are attention. The other 52 are Mamba layers with parameters `in_proj`, `out_proj`, `A_log`, `D` — none of which were targeted by LoRA.

## Root Cause

1. PEFT PR #3289 added `nemotron_h` to `TRANSFORMERS_MODELS_TO_LORA_TARGET_MODULES_MAPPING` with only attention modules
2. PEFT PR #2562 added `nemotron_h` to `mamba_model_types` in `_check_lora_target_modules_mamba()`, which **forbids** targeting Mamba modules (`in_proj`, `out_proj`, `conv1d`)
3. The combination means: PEFT defaults only target attention layers, and PEFT actively blocks you from targeting Mamba layers
4. No warning is emitted when target modules match an unusually low number of parameters

### Fix 1: Warn on low coverage

When `LoraConfig.target_modules` matches fewer than N% of linear layers in the model, emit a warning:

```python
if matched_layers / total_linear_layers < 0.3:
    warnings.warn(
        f"LoRA target modules only matched {matched_layers}/{total_linear_layers} "
        f"linear layers ({100*matched/total:.1f}%). For hybrid architectures like "
        f"NemotronH, consider including Mamba modules (in_proj, out_proj) in target_modules."
    )
```

### Fix 2: Document hybrid architecture target modules

Add documentation for hybrid Mamba+Attention models showing the correct target modules:

```python
# For NemotronH / Jamba / Granite-H hybrid models:
LoraConfig(
    target_modules=[
        # Attention layers
        "q_proj", "k_proj", "v_proj", "o_proj",
        # Mamba layers (requires removing from mamba_model_types check)
        "in_proj", "out_proj", "x_proj", "dt_proj",
    ],
    # Skip: conv1d, A_log, D (structured SSM parameters, not suitable for LoRA)
)
```

### Fix 3: Reconsider the Mamba module ban

PR #2562 banned Mamba modules from LoRA targeting due to shape issues. But `in_proj` and `out_proj` are standard `nn.Linear` layers — they should be LoRA-compatible. The ban should be scoped to `conv1d` and structured SSM parameters (`A_log`, `D`), not all Mamba modules.

## Environment

- peft: latest (with PR #3289 merged)
- transformers: 5.16.0.dev0
- Model: nvidia/NVIDIA-Nemotron-Nano-9B-v2 (56 layers, 4 attention + 52 Mamba)
- GPU: NVIDIA GB10 (119GB VRAM)

## Related

- PEFT PR #3289: Added default target_modules for nemotron_h (attention only)
- PEFT PR #2562: Added Mamba target module check (bans in_proj, out_proj)
- Unsloth issue #3854: "Trouble fine-tuning Nemotron 3 Nano, failure when merging LoRA"
- Full writeup: https://github.com/davidnichols-ops/claude-yolo-vibes-v5/blob/master/EVALUATION.md

**Comment by BenjaminBossan** [comment]
Thanks for reporting this. I agree that it's not good to add bad defaults, as users may use them without giving it another thought. It is, however, difficult to provide good default for LoRA targets (or other PEFT adapters for that matter) because the best target depends on the task. Just as an example, a task requiring low learning capacity and that is very sensitive to catastrophic forgetting might actually benefit from targeting only a few modules. There is no one size fits all solution.

That said, the fact that you had zero learning is very strange, it almost looks like something else is going on there. Did you check if any of the LoRA weights were updated at all?

As for the proposed fixes:

1. That's not an option, it's impossible for us to predetermine what would constitute a "low coverage", especially across different PEFT methods. Giving warnings unnecessarily has a cost, as it trains users to ignore PEFT warnings. In our docs and examples, we often show that after creating the PEFT model, you should call `model.print_trainable_parameters()`, which shows how many trainable parameters there are. If you think we can better document this, LMK.
2. We can add some information about good defaults for this type of model architecture. Did you try this setting and found it to work well?
3. #2562 is not forbidding `in_proj` AFAICT, where did you see that? For `out_proj` and `conv1d`, the reason why they're forbidden is _not_ because they use `Conv1D` layers, which are in fact supported. The reason is instead that the forward methods uses special kernels that would skip LoRA application, see the discussion in #2556.

**Comment by davidnichols-ops** [comment]
Thanks for the detailed response. You raise fair points — let me address each.

**On zero learning / LoRA weights:**

Yes, the LoRA weights were being updated — `print_trainable_parameters()` showed 18M trainable params (0.2% of the 9B model). The training ran for 25 steps and the optimizer was stepping. The issue was that DPO loss stayed at exactly 0.6931 (ln(2)) for all 25 steps, which is the initialization point — meaning the model's output distribution wasn't changing despite the weights updating. This makes sense: the 4 attention layers (q/k/v/o_proj) that LoRA targeted are only 7.1% of the 56 layers. The Mamba layers (52/56) carry most of the model's capacity, and they were frozen. The attention-only LoRA couldn't shift the output distribution enough to produce a DPO signal.

**On in_proj/out_proj:**

We did not re-run with `in_proj`/`out_proj` as targets — we ran out of GPU time on the MI300X. The v5 experiment is documented as a case study (the training failed, we identified the root cause, filed these issues, and moved back to Qwen 7B for production). So I can't claim these modules work well for LoRA on NemotronH — only that omitting them guarantees zero learning on a Mamba-heavy model.

**On the warning approach:**

You're right that a generic "low coverage" warning is hard to calibrate and could cause warning fatigue. I'll close PR #3556. 

A more targeted alternative: when `LoraConfig(target_modules=[...])` is created and **none** of the listed modules exist in the model, PEFT already handles this (silently skips). But what about the case where *some* exist and *some* don't? Currently there's no signal to the user that e.g. `gate_proj` was silently skipped because it doesn't exist in the model. 

Would a debug-level log (not a warning) when a listed target module doesn't match any layer in the model be acceptable? That way `print_trainable_parameters()` still shows the summary, but users debugging "why isn't my model learning?" can enable debug logging and see which modules were skipped.

**On #2562:**

You're right — I rechecked and #2562 adds `in_proj` to the forbidden list for Mamba models, not `out_proj`. The forbidden list I see in the code is for `out_proj` and `conv1d` on specific architectures. The reason (special kernels skipping LoRA application, per #2556) makes sense. I'll update the issue to correct this.

**On documentation:**

The most impactful fix is probably documentation. A note in the PEFT docs for hybrid architectures (Mamba, Jamba, NemotronH) saying "the majority of layers may be Mamba/SSM layers with different module names — check `model.named_modules()` to verify your target_modules match" would prevent the silent failure we hit. The TRL example fix (PR #6777) also helps since it's the official Nemotron 3 fine-tuning example.

Happy to contribute a docs PR if that direction sounds good.

**Comment by BenjaminBossan** [comment]
> The attention-only LoRA couldn't shift the output distribution enough to produce a DPO signal.

I see, so this is related to the DPO training then. Because normally, with SFT, even with only 0.2% trainable parameters, we should see weight updates and the loss moving.

> So I can't claim these modules work well for LoRA on NemotronH — only that omitting them guarantees zero learning on a Mamba-heavy model.

Thanks for letting me know. So it's not exactly like we can make a recommendation, as we don't have confirmed targets that work.

> A more targeted alternative: when `LoraConfig(target_modules=[...])` is created and **none** of the listed modules exist in the model, PEFT already handles this (silently skips). But what about the case where _some_ exist and _some_ don't? Currently there's no signal to the user that e.g. `gate_proj` was silently skipped because it doesn't exist in the model.

Yes, that has been bugging us for a while. It's not a trivial problem to solve: As a first step, we would need to determine which parts of `target_modules` was being used and which wasn't. Then we have multiple arguments that modify what is being targeted: besides `target_modules`, we have `exclude_modules`, `layers_to_transform`, `layers_pattern`, `target_parameters`. If a user passes `target_modules=["in_proj", "out_proj"]`, but no `out_proj` was targeted, but the reason is that `out_proj` was added to `exclude_modules`, how is that handled? And then we have an even more complicated case: `target_modules` can be a string, in which case it's treated as a regex match. How to determine which part of the regex was used?

Moreover, we need to cover all these edge cases with tests and also ensure backward compatibility. This is thus a very difficult problem to solve, which is why we haven't touched it yet. If you have a good idea, feel free to propose a PR, but it's not going to be easy :)

> The most impactful fix is probably documentation. A note in the PEFT docs for hybrid architectures (Mamba, Jamba, NemotronH) saying "the majority of layers may be Mamba/SSM layers with different module names — check `model.named_modules()` to verify your target_modules match" would prevent the silent failure we hit. The TRL example fix (PR #6777) also helps since it's the official Nemotron 3 fine-tuning example.
> 
> Happy to contribute a docs PR if that direction sounds good.

Yes, that sounds good to me. Probably the troubleshooting page is the best place to put this. Remember to word this in a PEFT method agnostic way, as this affects not only LoRA (but if you give an example, LoRA is fine).

**Comment by davidnichols-ops** [comment]
Thanks @BenjaminBossan, that's a helpful explanation of why the "silently skipped modules" problem is hard to solve generally — the interaction with `exclude_modules`, `layers_to_transform`, `layers_pattern`, `target_parameters`, and regex `target_modules` makes it a non-trivial change. I won't attempt that PR.

For the docs PR: I'll add a section to the troubleshooting page about hybrid architectures (Mamba, Jamba, NemotronH, etc.), worded PEFT-method-agnostically but with a LoRA example. The key points:
- Hybrid models may have layers with different module names than standard Transformer architectures
- `target_modules` that don't match any layers in the model are silently skipped
- Use `model.named_modules()` or `print_trainable_parameters()` to verify which modules are actually being targeted
- If trainable parameters are unexpectedly low, check that your `target_modules` match the actual module names in the model

I'll open that PR soon and link it here.

**Comment by Amarnath10i** [comment]
Following up on the docs direction @BenjaminBossan suggested (troubleshooting page, PEFT-method-agnostic with a LoRA example), I drafted the section and opened PR #3601. It explains that 	arget_modules entries not matching any module are silently skipped, how to verify via print_trainable_parameters() / 
amed_modules(), and calls out hybrid (attention + Mamba/SSM) architectures. @davidnichols-ops if you're still planning to open your own version, happy to hand this off or coordinate so we don't duplicate effort.

**Comment by Amarnath10i** [comment]
Correction: the PR is #3600 (not #3601). Apologies for the typo.

**Comment by davidnichols-ops** [comment]
@Amarnath10i thanks for the heads-up and for the initial draft in #3600. I had a clean docs branch sitting locally from the earlier discussion with @BenjaminBossan, so I rebased it on upstream/main and opened #3603. It covers the same silent-skip problem on hybrid (Mamba/SSM + attention) architectures and goes on the troubleshooting page as requested. Since #3600 is now closed, let's consolidate there. Happy to fold in any of the wording or examples you had in #3600 if you think they'd improve it.

**Comment by davidnichols-ops** [comment]
Quick wrap-up after checking the current upstream state across PEFT, TRL, and Transformers:

## What's in flight / done

- **PEFT docs:** #3603 adds a troubleshooting section for the silent `target_modules` skip on hybrid architectures (Mamba/Attention, e.g., NemotronH, Jamba). If it merges, the documentation side of this issue is covered.

## What still needs doing (post-merge)

- **PEFT default `target_modules`:** `TRANSFORMERS_MODELS_TO_LORA_TARGET_MODULES_MAPPING["nemotron_h"]` still defaults to `["q_proj", "k_proj", "v_proj", "o_proj"]`. `in_proj` is *not* in PEFT's Mamba forbidden list (`{"out_proj", "conv1d"}`), so adding `in_proj` to the default is the logical next step — but only after someone validates that training actually works with it. I don't have fresh GPU runs to confirm this, so I'm leaving it as a tracked follow-up rather than guessing.
- **TRL example:** `examples/sft_nemotron_3/sft_nemotron_3.py` still uses `--lora_target_modules q_proj k_proj v_proj o_proj gate_proj up_proj down_proj` and makes no mention of Mamba layers. It should include `in_proj` and ideally cross-reference the PEFT troubleshooting note once #3603 is live. `out_proj` should *not* be added because PEFT forbids it on Mamba-based architectures.

## Adjacent but not directly part of this fix

- **Transformers:** `NemotronHConfig.num_hidden_layers` is now exposed as a property derived from `layers_block_type` (commit 8436e1e). This is unrelated to the `target_modules` matching logic, but it's adjacent Nemotron-H tooling hygiene that helps GGUF/PEFT downstream tools see the correct layer count.

## What's deliberately out of scope for now

- A runtime warning for silently skipped `target_modules` names. As discussed earlier, this is entangled with `exclude_modules`, `layers_to_transform`, `layers_pattern`, regex `target_modules`, and `target_parameters`, so it's a larger change. The troubleshooting docs are the pragmatic fix.

Once #3603 merges, I think this issue can be closed from the PEFT side, with the default-target-modules and TRL-example items tracked as separate, validated follow-ups.

**Comment by Amarnath10i** [comment]
Thanks @davidnichols-ops (and @BenjaminBossan) — I've left #3600 closed and we can consolidate on your #3603 for the documentation side of this issue. Apologies for the preemptive PR; that's a mistake on my end and I'll avoid opening PRs when the original poster has already signaled they'll contribute.
